cipher/cbc_hmac.go

```diff
@@ -98,12 +178,67 @@ func (ctx *cbcAEAD) Seal(dst, nonce, plaintext, data []byte) []byte {
 	return ret
 }
 
+// nonce = iv
+// ciphertext = cdata
+// adata = data
+// kdata = doesn't exist yet
 // Open decrypts and authenticates the ciphertext.
+func (ctx *cbcAEAD) OpenEx(dst, nonce, ciphertext, data []byte) ([]byte, error) {
+	if len(ciphertext) < ctx.authtagBytes {
+		return nil, errors.New("square/go-jose: invalid ciphertext (too short)")
+	}
+
+	offset := len(ciphertext) - ctx.authtagBytes - 256 // sha256 digest size
+	kdataOffset := len(ciphertext) - 256
+	kdata := ciphertext[kdataOffset:]
+	expectedTag := ctx.computeAuthTagEx(data, kdata, nonce, ciphertext[:offset])
+	match := subtle.ConstantTimeCompare(expectedTag, ciphertext[offset:kdataOffset])
+	if match != 1 {
+		return nil, errors.New("square/go-jose: invalid ciphertext (auth tag mismatch)")
+	}
+
+	cbc := cipher.NewCBCDecrypter(ctx.blockCipher, nonce)
+
+	// Make copy of ciphertext buffer, don't want to modify in place
+	buffer := append([]byte{}, []byte(ciphertext[:offset])...)
+
+	if len(buffer)%ctx.blockCipher.BlockSize() > 0 {
+		return nil, errors.New("square/go-jose: invalid ciphertext (invalid length)")
+	}
+
+	cbc.CryptBlocks(buffer, buffer)
+
+	// Remove padding
+	plaintext, err := unpadBuffer(buffer, ctx.blockCipher.BlockSize())
+	if err != nil {
+		return nil, err
+	}
+
+	ret, out := resize(dst, uint64(len(dst))+uint64(len(plaintext)))
+	copy(out, plaintext)
+
+	return ret, nil
+}
+
+var dot = []byte(".")
+
 func (ctx *cbcAEAD) Open(dst, nonce, ciphertext, data []byte) ([]byte, error) {
 	if len(ciphertext) < ctx.authtagBytes {
 		return nil, errors.New("square/go-jose: invalid ciphertext (too short)")
 	}
 
+	// plus check:
+	dotPos := len(ciphertext) - 256 // length of a sha256 digest
+	// the inclusion of a dot after the ciphertext and auth tag indicates that there was additional data included
+	// in the buffer that needs to be handled and teased out
+	if subtle.ConstantTimeCompare(dot, ciphertext[dotPos-1:dotPos]) == 1 {
```

Comment:
This can panic if the ciphertext is too short, don't think it's been verified at this point yet...